# Финальны проект. SQL

**Описание проекта:** Компания купила крупный сервис для чтения книг по подписке. Наша задача проанализировать базу данных.
Дать ответы на вопросы: 
- Сколько книг вышло после 1 января 2000 года;
- Количество обзоров и средняя оценка для каждой книги;
- Издательство, которое выпустило наибольшее число книг толще 50 страниц;
- Автор с самой высокой средней оценкой книг (только книги с 50 и более оценками);
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Описаные данных:** 

**Таблица 'books'**
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**
- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Откроем даныые

In [53]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Таблица books

In [129]:
query = '''
           SELECT *
           FROM books
        '''

books = pd.io.sql.read_sql(query, con = engine)
books.to_csv('books.csv', index=False)
books.head(5)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

## Таблица authors

In [55]:
query = '''
           SELECT *
           FROM authors
        '''

authors = pd.io.sql.read_sql(query, con = engine)
authors.to_csv('authors.csv', index=False)
authors.head(5)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

## Таблица publishers

In [56]:
query = '''
           SELECT *
           FROM publishers
        '''

publishers = pd.io.sql.read_sql(query, con = engine)
publishers.to_csv('publishers.csv', index=False)
publishers.head(5)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

## Таблица ratings

In [57]:
query = '''
           SELECT *
           FROM ratings
        '''

ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head(5)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

## Таблица reviews

In [58]:
query = '''
           SELECT *
           FROM reviews
        '''

reviews = pd.io.sql.read_sql(query, con = engine)
reviews.to_csv('reviews.csv', index=False)
reviews.head(5)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## Решение задач 

### Cколько книг вышло после 1 января 2000 года

In [59]:
query = '''
           SELECT COUNT(book_id)
           FROM books
           WHERE publication_date >= '2000-01-01'
        '''

count_book = pd.io.sql.read_sql(query, con = engine)
count_book

count
0    821

*С 1 января 2000 года вышло 821 книг*

### Количество обзоров и средняя оценка для каждой книги  

In [60]:
query = '''
           WITH 
           r AS (
               SELECT book_id, count(review_id)
               FROM reviews
               GROUP BY 1
               ),
           b AS (
                SELECT *
                FROM books
                ),
           i AS (
                SELECT book_id, AVG(rating)
                FROM ratings
                GROUP BY 1)
            
            SELECT b.book_id, b.title, r.count, i.avg
            FROM b
            LEFT JOIN r ON b.book_id = r.book_id
            LEFT JOIN i ON b.book_id = i.book_id
            
            
            
            
        '''

count_rev = pd.io.sql.read_sql(query, con = engine)
count_rev.head()

book_id                                              title  count       avg
0        1                                       'Salem's Lot    2.0  3.666667
1        2                 1 000 Places to See Before You Die    1.0  2.500000
2        3  13 Little Blue Envelopes (Little Blue Envelope...    3.0  4.666667
3        4  1491: New Revelations of the Americas Before C...    2.0  4.500000
4        5                                               1776    4.0  4.000000

### Издательство, которое выпустило наибольшее число книг толще 50 страниц 

In [67]:
query = '''         
            
            SELECT publisher, COUNT(title)
           FROM books as b
           LEFT JOIN publishers AS pu ON b.publisher_id  = pu.publisher_id
           WHERE num_pages > 50 
           GROUP BY publisher
           ORDER BY count DESC
           LIMIT 1
                
            
        '''

fat = pd.io.sql.read_sql(query, con = engine)
fat.head()

publisher  count
0  Penguin Books     42

*Больше всего книг толще 50 страниц выпустило издательство Penguin Books, выпустившее 42 книги*

### Автор с самой высокой средней оценкой книг (только книги с 50 и более оценками)

In [135]:
query = '''         
            SELECT author, AVG(t.avg)
            FROM (SELECT title, COUNT(rating), AVG(rating)
                FROM books as b
                LEFT JOIN ratings AS r ON b.book_id  = r.book_id
                GROUP BY title) AS t
                LEFT JOIN books AS b ON t.title=b.title
                LEFT JOIN authors AS a ON b.author_id=a.author_id
                WHERE t.count >= 50
                GROUP BY author
                ORDER BY avg DESC
                LIMIT 1
                   
            
        '''

top = pd.io.sql.read_sql(query, con = engine)
top.head()

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

*Автором с самой высокой средней оценкой с более 50 оценками является J.K. Rowling/Mary GrandPré, у него рейтинг 4.283844*

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [110]:
query = '''         
            
             
            SELECT AVG(count)
            FROM (
                SELECT r.username, COUNT(rating)
                FROM (
                    SELECT r.username, COUNT(rating_id) as rating
                    FROM books AS b
                    LEFT JOIN ratings AS r ON b.book_id  = r.book_id
                    GROUP BY r.username
                ) AS u
                RIGHT JOIN reviews AS r ON u.username  = r.username
                WHERE rating > 50
                GROUP BY r.username
                ) AS t
            
            
            
                           
            
        '''

top = pd.io.sql.read_sql(query, con = engine)
top.head()

avg
0  24.333333

*Среднее количество обзоров от пользователей с более 50 оценками 24,33*

## Вывод

В ходе данной работы, мы проанализировали базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Выяснили, что с 1 января 2000 года вышло 821 книг. Больше всего книг толще 50 страниц выпустило издательство Penguin Books, выпустившее 42 книги. Автором с самой высокой средней оценкой с более 50 оценками является J.K. Rowling/Mary GrandPré, у него рейтинг 4.283844. Среднее количество обзоров от пользователей с более 50 оценками 24,33